### DISCUSSION TITLE
Human Demos for NodeJS

### DISCUSSION THREAD

**Discussion by saeidn95** [vladmandic/human#442] [Q&A]
I am referring to the instruction in (https://github.com/vladmandic/human/tree/main/demo/nodejs) for running the Human Demos for NodeJS. 

I have tried (node demo/nodejs/node.js) from a terminal. However, I get the following error message. I wonder if it a version compatibility issue or something else. I am using v12.22.9 on Ubuntu 22.04.4 LTS

~/human/demo/nodejs/node.js:68
  log.data('Image bytes:', buffer?.length, 'buffer:', buffer?.slice(0, 32));
                                  ^

SyntaxError: Unexpected token '.'

  at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)
    at internal/main/run_main_module.js:17:47

**Comment by vladmandic** [comment]
`package.json` does specify:

```json
  "engines": {
    "node": ">=14.0.0"
  },
```

nodejs 12 was introduced in 2019-04-23 and had end-of-life in 2022-04-30 which is 2 years ago.

**Reply by saeidn95** [reply]
Thank you for the help.

I have upgrade Node.js to v20.12.1. Now I get the following error message complaining about not finding  module '@vladmandic/pilogger' Do I have to download some additional modules?

node ./demo/nodejs/node.js 
node:internal/modules/cjs/loader:1146
  throw err;
  ^

Error: Cannot find module '@vladmandic/pilogger'
Require stack:
- /human/demo/nodejs/node.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1143:15)
    at Module._load (node:internal/modules/cjs/loader:984:27)
    at Module.require (node:internal/modules/cjs/loader:1231:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (~/human/demo/nodejs/node.js:8:13)
    at Module._compile (node:internal/modules/cjs/loader:1369:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1427:10)
    at Module.load (node:internal/modules/cjs/loader:1206:32)
    at Module._load (node:internal/modules/cjs/loader:1022:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/home/snooshab/work/DMS/collections/human/demo/nodejs/node.js' ]
}

Node.js v20.12.1

**Reply by vladmandic** [reply]
to run demos, you need to install dependencies.
`npm install`

**Reply by saeidn95** [reply]
I ran "npm install" from the human folder I also ran "sudo npm install". Neither installed the vladmandic/pilogger' dependency I still get the Error: Cannot find module '@vladmandic/pilogger'

Do I have to clone the git repo with some options?

**Reply by saeidn95** [reply]
Update:   I also tried rebuild the prohect by following the instruction (https://github.com/vladmandic/human/wiki/Build-Process)
npm install --dev # installs all project dependencies
npm run build
I still get the same error "Error: Cannot find module '@vladmandic/pilogger'"

Next I commented out const "log = require('@vladmandic/pilogger')" from node.js file. Next I get a new error "Error: Cannot find module '@tensorflow/tfjs-node'"

I tried to install with "  npm install @vladmandic/human @tensorflow/tfjs"
as stated in (https://github.com/vladmandic/human/wiki/Install). This did not work either I still get the same error. Some how the dependencies do not get installed or something.




npm install @vladmandic/human @tensorflow/tfjs

**Comment by vladmandic** [comment]
git clone https://github.com/vladmandic/human human-test
    cd human-test
    npm install --dev
    node demo/nodejs/node-simple.js ~/generative/Samples/cutie-512.png

```log
12:02:50.907 Human: version: 3.2.1
12:02:50.908 Human: tfjs version: 4.17.0
12:02:50.908 Human: environment: {
  browser: false,
  node: true,
  platform: '',
  agent: '',
  backends: [],
  tfjs: { version: '4.17.0' },
  offscreen: false,
  tensorflow: {},
  wasm: {},
  webgl: {},
  webgpu: {},
  cpu: { flags: [] }
}
Human: 3.2.1 TF: 4.17.0
12:02:50.910 Human: available backends: [ 'cpu', 'tensorflow' ]
12:02:50.910 Human: setting backend: [ 'tensorflow' ]
12:02:50.911 Human: registered kernels: [ 'floormod' ]
12:02:50.924 Human: load: {
  model: 'blazeface',
  url: 'file://models/blazeface.json',
  bytes: 538928
}
12:02:50.926 Human: load: { model: 'emotion', url: 'file://models/emotion.json', bytes: 820516 }
12:02:50.934 Human: load: {
  model: 'facemesh',
  url: 'file://models/facemesh.json',
  bytes: 1477958
}
12:02:50.941 Human: load: {
  model: 'handlandmark-lite',
  url: 'file://models/handlandmark-lite.json',
  bytes: 2023432
}
12:02:50.945 Human: load: { model: 'iris', url: 'file://models/iris.json', bytes: 2599092 }
12:02:50.969 Human: load: {
  model: 'handtrack',
  url: 'file://models/handtrack.json',
  bytes: 2964837
}
12:02:50.983 Human: load: { model: 'faceres', url: 'file://models/faceres.json', bytes: 6978814 }
12:02:50.983 Human: tf engine state: 36772028 bytes 1304 tensors
12:02:51.279 Human: warmup full 296 ms
loaded input file: /home/vlado/generative/Samples/cutie-512.png resolution: [ 512, 512, 3 ]
{
  face: [
    {
      id: 0,
      mesh: [Array],
      meshRaw: [Array],
      box: [Array],
      boxRaw: [Array],
      score: 1,
      boxScore: 0.95,
      faceScore: 1,
      size: [Array],
      annotations: [Object],
      age: 16.3,
      gender: 'female',
      genderScore: 0.79,
      embedding: [Array],
      emotion: [Array],
      distance: 0.86,
      rotation: [Object]
    }
  ],
  body: [],
  hand: [],
  gesture: [
    { face: 0, gesture: 'facing right' },
    { face: 0, gesture: 'head down' },
    { iris: 0, gesture: 'facing center' },
    { iris: 0, gesture: 'looking up' }
  ],
  object: [],
  performance: {
    initBackend: 1,
    loadModels: 74,
    inputProcess: 0,
    totalFrames: 2,
    cachedFrames: 0,
    cacheCheck: 0,
    total: 188,
    warmup: 295
  },
  canvas: null,
  timestamp: 1712764971482,
  error: null,
  width: 512,
  height: 512,
  persons: [Getter]
}
```

**Reply by saeidn95** [reply]
OMM! Great! it works now! I swear I followed the same steps before!
Thanks a lot. I can also build using "npm run build"
